File: src/stores/repo.ts
import { defineStore } from 'pinia';
import { ref } from 'vue';
import type { RepositoryInfo, RepoStatusSummary, FileStatusItem, BranchItem, RemoteItem, TagItem, StashItem, SyncStatus } from '@/types/git';
import type { GraphCommitNode } from '@/types/graph';
import { formatGitError, useGitApi } from '@/composables/useGitApi';
import { useConsoleStore } from '@/stores/console';
import { useDiffStore } from '@/stores/diff';
import { CONFIG_KEYS, persistAppConfig } from '@/services/appConfig';

export interface ManagedRepo {
  path: string;
  name: string;
  lastOpened: number;
}

function normalizeStoredRepoPath(path: string): string {
  if (path.startsWith('\\\\?\\UNC\\')) return `\\\\${path.slice(8)}`;
  if (path.startsWith('\\\\?\\')) return path.slice(4);
  return path;
}

function emptyStatusSummary(): RepoStatusSummary {
  return {
    staged_files: [],
    unstaged_files: [],
    untracked_files: [],
    conflicted_files: [],
    total_changes: 0,
  };
}

function getInitialRepos(): ManagedRepo[] {
  try {
    const raw = localStorage.getItem(CONFIG_KEYS.repositories);
    if (raw) {
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed) && parsed.length > 0) {
        return parsed.map((repo) => ({ ...repo, path: normalizeStoredRepoPath(repo.path) }));
      }
    }
  } catch {}
  return [];
}

function getInitialActiveRepo(): string {
  try {
    const saved = localStorage.getItem(CONFIG_KEYS.activeRepository);
    if (saved) return normalizeStoredRepoPath(saved);
  } catch {}
  return '';
}

export const useRepoStore = defineStore('repo', () => {
  const gitApi = useGitApi();
  const consoleStore = useConsoleStore();
  const diffStore = useDiffStore();
  const repoList = ref<ManagedRepo[]>(getInitialRepos());
  const activeRepoPath = ref<string>(getInitialActiveRepo());
  const repoInfo = ref<RepositoryInfo | null>(null);
  const statusSummary = ref<RepoStatusSummary>(emptyStatusSummary());
  const branches = ref<BranchItem[]>([]);
  const remotes = ref<RemoteItem[]>([]);
  const tags = ref<TagItem[]>([]);
  const stashes = ref<StashItem[]>([]);
  const syncStatus = ref<SyncStatus>({ incoming: [], outgoing: [] });
  const isSyncStatusOpen = ref(false);
  const isWorktreeManagerOpen = ref(false);
  const isPullRequestOpen = ref(false);
  const commitNodes = ref<GraphCommitNode[]>([]);
  const graphHasMore = ref(false);
  const isLoadingMoreCommits = ref(false);
  const selectedCommit = ref<GraphCommitNode | null>(null);
  const selectedCommitFiles = ref<FileStatusItem[]>([]);
  const branchComparison = ref<{
    baseBranch: string;
    targetBranch: string;
    baseCommitId: string;
    targetCommitId: string;
  } | null>(null);
  const branchComparisonFiles = ref<FileStatusItem[]>([]);
  const isLoading = ref<boolean>(false);
  const errorMessage = ref<string | null>(null);

  // Dialog & Modal states
  const isAddRepoModalOpen = ref<boolean>(false);
  const isBranchModalOpen = ref<boolean>(false);
  const isTagModalOpen = ref<boolean>(false);
  const isStashModalOpen = ref<boolean>(false);
  const isMergeModalOpen = ref<boolean>(false);
  const isRebaseModalOpen = ref<boolean>(false);
  const isRenameBranchModalOpen = ref<boolean>(false);
  const isResetModalOpen = ref<boolean>(false);
  const isRemoteModalOpen = ref<boolean>(false);

  // Context target data
  const targetBranchForAction = ref<string>('');
  const targetCommitForAction = ref<string>('');

  const saveReposToStorage = () => {
    try {
      localStorage.setItem(CONFIG_KEYS.repositories, JSON.stringify(repoList.value));
      localStorage.setItem(CONFIG_KEYS.activeRepository, activeRepoPath.value);
    } catch {}
    void persistAppConfig().catch((error) => {
      consoleStore.logWarning('Failed to save user configuration.', formatGitError(error));
    });
  };

  const clearLoadedRepoData = () => {
    repoInfo.value = null;
    statusSummary.value = emptyStatusSummary();
    branches.value = [];
    remotes.value = [];
    tags.value = [];
    stashes.value = [];
    commitNodes.value = [];
    graphHasMore.value = false;
    selectedCommit.value = null;
    selectedCommitFiles.value = [];
    branchComparison.value = null;
    branchComparisonFiles.value = [];
  };

  const loadRepo = async (targetPath?: string) => {
    const path = targetPath || activeRepoPath.value;
    if (!path) {
      clearLoadedRepoData();
      return false;
    }
    if (path !== activeRepoPath.value) {
      diffStore.clearSelection();
    }
    isLoading.value = true;
    errorMessage.value = null;
    activeRepoPath.value = path;
    saveReposToStorage();
    clearLoadedRepoData();

    try {
      const info = await gitApi.getRepoInfo(path);
      repoInfo.value = info;

      const existing = repoList.value.find((r) => r.path === path);
      if (existing && info.name) {
        existing.name = info.name;
        existing.lastOpened = Date.now();
        saveReposToStorage();
      }

      const [statusResult, branchResult, remoteResult, tagResult, stashResult, graphResult] = await Promise.allSettled([
        gitApi.getRepoStatus(path),
        gitApi.listBranches(path),
        gitApi.listRemotes(path),
        gitApi.listTags(path),
        gitApi.listStashes(path),
        gitApi.getCommitGraph(path, 0, 150),
      ]);

      const failures: string[] = [];
      if (statusResult.status === 'fulfilled') statusSummary.value = statusResult.value;
      else failures.push(`status: ${formatGitError(statusResult.reason)}`);
      if (branchResult.status === 'fulfilled') branches.value = branchResult.value;
      else failures.push(`branches: ${formatGitError(branchResult.reason)}`);
      if (remoteResult.status === 'fulfilled') remotes.value = remoteResult.value;
      else failures.push(`remotes: ${formatGitError(remoteResult.reason)}`);
      if (tagResult.status === 'fulfilled') tags.value = tagResult.value;
      else failures.push(`tags: ${formatGitError(tagResult.reason)}`);
      if (stashResult.status === 'fulfilled') stashes.value = stashResult.value;
      else failures.push(`stashes: ${formatGitError(stashResult.reason)}`);
      if (graphResult.status === 'fulfilled') {
        commitNodes.value = graphResult.value.nodes;
        graphHasMore.value = graphResult.value.has_more;
        selectedCommit.value = graphResult.value.nodes[0] || null;
      } else {
        failures.push(`commit graph: ${formatGitError(graphResult.reason)}`);
      }

      if (failures.length) {
        errorMessage.value = failures.join('; ');
        consoleStore.logWarning('Repository opened with partial data.', errorMessage.value);
      }
      return true;
    } catch (err: any) {
      clearLoadedRepoData();
      errorMessage.value = formatGitError(err, 'Failed to open repository');
      consoleStore.logError('Failed to open repository.', errorMessage.value);
      return false;
    } finally {
      isLoading.value = false;
    }
  };

  const addRepo = async (newPath: string) => {
    const validation = await gitApi.validateRepo(newPath);
    if (!validation.valid) {
      throw new Error(validation.message || 'Invalid Git repository');
    }
    const cleanPath = normalizeStoredRepoPath(validation.path || newPath);
    const name = validation.name || cleanPath.split(/[\\/]/).pop() || 'Repo';

    const idx = repoList.value.findIndex((r) => r.path === cleanPath);
    if (idx === -1) {
      repoList.value.push({ path: cleanPath, name, lastOpened: Date.now() });
    } else {
      repoList.value[idx].lastOpened = Date.now();
    }
    saveReposToStorage();
    const loaded = await loadRepo(cleanPath);
    if (!loaded) throw new Error(errorMessage.value || 'Failed to open repository');
  };

  const removeRepo = (pathToRemove: string) => {
    repoList.value = repoList.value.filter((r) => r.path !== pathToRemove);
    if (activeRepoPath.value === pathToRemove) {
      if (repoList.value.length > 0) {
        loadRepo(repoList.value[0].path);
      } else {
        activeRepoPath.value = '';
        clearLoadedRepoData();
      }
    }
    saveReposToStorage();
  };

  const switchRepo = async (path: string) => {
    await loadRepo(path);
  };

  const stageFile = async (filePath: string) => {
    await gitApi.stageFile(activeRepoPath.value, filePath);
    await loadRepo(activeRepoPath.value);
  };

  const stageAll = async () => {
    await gitApi.stageAll(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const unstageFile = async (filePath: string) => {
    await gitApi.unstageFile(activeRepoPath.value, filePath);
    await loadRepo(activeRepoPath.value);
  };

  const unstageAll = async () => {
    await gitApi.unstageAll(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const discardFile = async (filePath?: string) => {
    await gitApi.discardFile(activeRepoPath.value, filePath);
    await loadRepo(activeRepoPath.value);
  };

  const commit = async (message: string, author: string, email: string, options: { amend?: boolean; sign?: boolean; preCommitCommand?: string } = {}) => {
    await gitApi.createCommit(activeRepoPath.value, message, author, email, options);
    await loadRepo(activeRepoPath.value);
  };

  const commitAndPush = async (message: string, author: string, email: string) => {
    await gitApi.commitAndPush(activeRepoPath.value, message, author, email);
    await loadRepo(activeRepoPath.value);
  };

  const checkoutBranch = async (branchName: string) => {
    await gitApi.checkoutBranch(activeRepoPath.value, branchName);
    await loadRepo(activeRepoPath.value);
  };

  const createBranch = async (name: string, startPoint?: string, checkout = true) => {
    await gitApi.createBranch(activeRepoPath.value, name, startPoint, checkout);
    await loadRepo(activeRepoPath.value);
  };

  const deleteBranch = async (name: string, force = false) => {
    await gitApi.deleteBranch(activeRepoPath.value, name, force);
    await loadRepo(activeRepoPath.value);
  };

  const renameBranch = async (oldName: string, newName: string) => {
    await gitApi.renameBranch(activeRepoPath.value, oldName, newName);
    await loadRepo(activeRepoPath.value);
  };

  const updateRemoteUrl = async (remoteName: string, url: string, pushUrl?: string) => {
    await gitApi.setRemoteUrl(activeRepoPath.value, remoteName, url, pushUrl);
    remotes.value = await gitApi.listRemotes(activeRepoPath.value);
    repoInfo.value = await gitApi.getRepoInfo(activeRepoPath.value);
  };

  const createTag = async (name: string, message?: string, commitId?: string) => {
    await gitApi.createTag(activeRepoPath.value, name, message, commitId);
    await loadRepo(activeRepoPath.value);
  };

  const createStash = async (message?: string) => {
    await gitApi.createStash(activeRepoPath.value, message);
    await loadRepo(activeRepoPath.value);
  };

  const popStash = async (index = 0) => {
    await gitApi.popStash(activeRepoPath.value, index);
    await loadRepo(activeRepoPath.value);
  };

  // Merge, Rebase, Cherry-pick, Revert, Reset
  const mergeBranch = async (target: string, strategy: 'default' | 'no-ff' | 'squash' | 'ff-only' = 'default', message?: string) => {
    const res = await gitApi.mergeBranch(activeRepoPath.value, target, strategy, message);
    await loadRepo(activeRepoPath.value);
    return res;
  };

  const abortMerge = async () => {
    await gitApi.abortMerge(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const continueMerge = async () => {
    await gitApi.continueMerge(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const rebase = async (upstream: string) => {
    const res = await gitApi.rebase(activeRepoPath.value, upstream);
    await loadRepo(activeRepoPath.value);
    return res;
  };

  const continueRebase = async () => {
    await gitApi.continueRebase(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const abortRebase = async () => {
    await gitApi.abortRebase(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const cherryPick = async (commitId: string) => {
    const res = await gitApi.cherryPick(activeRepoPath.value, commitId);
    await loadRepo(activeRepoPath.value);
    return res;
  };

  const continueCherryPick = async () => {
    await gitApi.continueCherryPick(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const abortCherryPick = async () => {
    await gitApi.abortCherryPick(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const revertCommit = async (commitId: string) => {
    const res = await gitApi.revertCommit(activeRepoPath.value, commitId);
    await loadRepo(activeRepoPath.value);
    return res;
  };

  const selectCommit = async (commit: GraphCommitNode | null) => {
    branchComparison.value = null;
    branchComparisonFiles.value = [];
    selectedCommit.value = commit;
    if (!commit || !activeRepoPath.value) {
      selectedCommitFiles.value = [];
      diffStore.clearSelection();
      return;
    }
    try {
      const files = await gitApi.getCommitChanges(activeRepoPath.value, commit.id);
      selectedCommitFiles.value = files;
      if (files.length > 0) {
        await diffStore.selectFile(files[0].path, false, activeRepoPath.value, commit.id);
      } else {
        diffStore.clearSelection();
      }
    } catch {
      selectedCommitFiles.value = [];
      diffStore.clearSelection();
    }
  };

  const interactiveRebase = async (upstream: string, plan: import('@/types/git').RebasePlanItem[]) => {
    await gitApi.interactiveRebase(activeRepoPath.value, upstream, plan);
    await loadRepo(activeRepoPath.value);
  };

  const applyStash = async (index: number) => {
    await gitApi.applyStash(activeRepoPath.value, index);
    await loadRepo(activeRepoPath.value);
  };

  const dropStash = async (index: number) => {
    await gitApi.dropStash(activeRepoPath.value, index);
    await loadRepo(activeRepoPath.value);
  };

  const renameStash = async (index: number, message: string) => {
    await gitApi.renameStash(activeRepoPath.value, index, message);
    await loadRepo(activeRepoPath.value);
  };

  const createShelf = async (message: string, filePaths: string[]) => {
    await gitApi.createShelf(activeRepoPath.value, message, filePaths);
    await loadRepo(activeRepoPath.value);
  };

  const loadMoreCommits = async () => {
    if (!activeRepoPath.value || !graphHasMore.value || isLoadingMoreCommits.value) return [];
    isLoadingMoreCommits.value = true;
    try {
      const page = await gitApi.getCommitGraph(activeRepoPath.value, commitNodes.value.length, 150);
      const existing = new Set(commitNodes.value.map((commit) => commit.id));
      const additions = page.nodes.filter((commit) => !existing.has(commit.id));
      commitNodes.value.push(...additions);
      graphHasMore.value = page.has_more;
      return additions;
    } finally {
      isLoadingMoreCommits.value = false;
    }
  };

  const locateCommit = async (commitId: string) => {
    let commit = commitNodes.value.find((item) => item.id === commitId);
    while (!commit && graphHasMore.value) {
      const additions = await loadMoreCommits();
      commit = additions.find((item) => item.id === commitId);
      if (additions.length === 0 && graphHasMore.value) break;
    }
    if (!commit) return false;
    await selectCommit(commit);
    return true;
  };

  const locateRevision = async (revision: string) => {
    const commitId = await gitApi.resolveRevision(activeRepoPath.value, revision);
    return await locateCommit(commitId);
  };

  const compareBranch = async (branch: BranchItem) => {
    const baseCommitId = repoInfo.value?.head_commit_id;
    const baseBranch = repoInfo.value?.head_branch || 'HEAD';
    if (!activeRepoPath.value || !baseCommitId) {
      throw new Error('The current branch has no commit to compare.');
    }

    selectedCommit.value = null;
    selectedCommitFiles.value = [];
    diffStore.clearSelection();
    branchComparison.value = {
      baseBranch,
      targetBranch: branch.name,
      baseCommitId,
      targetCommitId: branch.target_commit_id,
    };

    try {
      const files = await gitApi.getBranchChanges(
        activeRepoPath.value,
        baseCommitId,
        branch.target_commit_id,
      );
      branchComparisonFiles.value = files;
      if (files.length > 0) {
        const first = files[0];
        await diffStore.selectBranchComparisonFile(
          first.path,
          first.old_path,
          activeRepoPath.value,
          baseCommitId,
          branch.target_commit_id,
        );
      }
      return files;
    } catch (error) {
      branchComparison.value = null;
      branchComparisonFiles.value = [];
      diffStore.clearSelection();
      throw error;
    }
  };

  const clearBranchComparison = () => {
    branchComparison.value = null;
    branchComparisonFiles.value = [];
    diffStore.clearSelection();
  };

  const continueRevert = async () => {
    await gitApi.continueRevert(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const abortRevert = async () => {
    await gitApi.abortMerge(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
  };

  const reset = async (target: string, mode: '--soft' | '--mixed' | '--hard' = '--mixed') => {
    await gitApi.reset(activeRepoPath.value, target, mode);
    await loadRepo(activeRepoPath.value);
  };

  const fetchRemote = async () => {
    await gitApi.fetchRemote(activeRepoPath.value);
    await loadRepo(activeRepoPath.value);
    syncStatus.value = await gitApi.getSyncStatus(activeRepoPath.value).catch(() => ({ incoming: [], outgoing: [] }));
  };

  const pullRemote = async (strategy: 'merge' | 'rebase' | 'ff-only' = 'merge') => {
    try {
      await gitApi.pullRemote(activeRepoPath.value, strategy);
    } finally {
      await loadRepo(activeRepoPath.value);
    }
  };

  const pushRemote = async (forceWithLease = false) => {
    await gitApi.pushRemote(activeRepoPath.value, forceWithLease);
    await loadRepo(activeRepoPath.value);
    syncStatus.value = await gitApi.getSyncStatus(activeRepoPath.value).catch(() => ({ incoming: [], outgoing: [] }));
  };

  const refreshSyncStatus = async () => {
    syncStatus.value = await gitApi.getSyncStatus(activeRepoPath.value);
  };

  const discoverRoots = async () => {
    const roots = await gitApi.discoverGitRoots(activeRepoPath.value);
    for (const path of roots) {
      if (repoList.value.some((repo) => repo.path === path)) continue;
      const validation = await gitApi.validateRepo(path);
      if (validation.valid) repoList.value.push({ path, name: validation.name || path.split(/[\\/]/).pop() || path, lastOpened: Date.now() });
    }
    saveReposToStorage();
    return roots.length;
  };

  return {
    repoList,
    activeRepoPath,
    repoInfo,
    statusSummary,
    branches,
    remotes,
    tags,
    stashes,
    syncStatus,
    isSyncStatusOpen,
    isWorktreeManagerOpen,
    isPullRequestOpen,
    commitNodes,
    graphHasMore,
    isLoadingMoreCommits,
    selectedCommit,
    selectedCommitFiles,
    branchComparison,
    branchComparisonFiles,
    isLoading,
    errorMessage,
    isAddRepoModalOpen,
    isBranchModalOpen,
    isTagModalOpen,
    isStashModalOpen,
    isMergeModalOpen,
    isRebaseModalOpen,
    isRenameBranchModalOpen,
    isResetModalOpen,
    isRemoteModalOpen,
    targetBranchForAction,
    targetCommitForAction,
    loadRepo,
    addRepo,
    removeRepo,
    switchRepo,
    stageFile,
    stageAll,
    unstageFile,
    unstageAll,
    discardFile,
    commit,
    commitAndPush,
    checkoutBranch,
    createBranch,
    deleteBranch,
    renameBranch,
    updateRemoteUrl,
    createTag,
    createStash,
    popStash,
    applyStash,
    dropStash,
    renameStash,
    createShelf,
    selectCommit,
    loadMoreCommits,
    locateCommit,
    locateRevision,
    compareBranch,
    clearBranchComparison,
    mergeBranch,
    abortMerge,
    continueMerge,
    rebase,
    interactiveRebase,
    continueRebase,
    abortRebase,
    cherryPick,
    continueCherryPick,
    abortCherryPick,
    revertCommit,
    continueRevert,
    abortRevert,
    reset,
    fetchRemote,
    pullRemote,
    pushRemote,
    refreshSyncStatus,
    discoverRoots,
  };
});

File: src-tauri/src/commands/repo.rs
use gitbx_core::{
    BranchItem, GitService, RemoteItem, RepoStatusSummary, RepositoryInfo, StashItem, TagItem,
};

type CommandResult<T> = std::result::Result<T, String>;

#[tauri::command]
pub async fn init_repo(repo_path: String) -> CommandResult<RepositoryInfo> {
    let repo = git2::Repository::init(&repo_path).map_err(|e| e.to_string())?;
    let path = repo
        .path()
        .parent()
        .unwrap_or(repo.path())
        .to_string_lossy()
        .to_string();
    GitService::info(&path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn clone_repo(url: String, destination: String) -> CommandResult<RepositoryInfo> {
    GitService::clone_repo(&url, &destination).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_repo_info(repo_path: String) -> CommandResult<RepositoryInfo> {
    GitService::info(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_repo_status(repo_path: String) -> CommandResult<RepoStatusSummary> {
    GitService::open(&repo_path)
        .map_err(|e| e.to_string())?
        .get_status()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_branches(repo_path: String) -> CommandResult<Vec<BranchItem>> {
    GitService::open(&repo_path)
        .map_err(|e| e.to_string())?
        .list_branches(None)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_remotes(repo_path: String) -> CommandResult<Vec<RemoteItem>> {
    GitService::open(&repo_path)
        .map_err(|e| e.to_string())?
        .list_remotes()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn set_remote_url(
    repo_path: String,
    remote_name: String,
    url: String,
    push_url: Option<String>,
) -> CommandResult<()> {
    GitService::set_remote_urls(&repo_path, &remote_name, &url, push_url.as_deref())
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn stage_file(repo_path: String, file_path: String) -> CommandResult<()> {
    GitService::with_write_lock(&repo_path, |repo| {
        GitService::validate_file_path(&repo_path, &file_path)?;
        repo.stage_file(&file_path)
    })
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn unstage_file(repo_path: String, file_path: String) -> CommandResult<()> {
    GitService::validate_file_path(&repo_path, &file_path).map_err(|e| e.to_string())?;
    GitService::with_write_lock(&repo_path, |repo| repo.unstage_file(&file_path))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn stage_all(repo_path: String) -> CommandResult<()> {
    GitService::with_write_lock(&repo_path, |repo| repo.stage_all()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn unstage_all(repo_path: String) -> CommandResult<()> {
    GitService::with_write_lock(&repo_path, |repo| repo.unstage_all()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn discard_file(repo_path: String, file_path: Option<String>) -> CommandResult<()> {
    GitService::discard_file(&repo_path, file_path.as_deref()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_commit(
    repo_path: String,
    message: String,
    author: String,
    email: String,
) -> CommandResult<String> {
    GitService::with_write_lock(&repo_path, |repo| {
        repo.create_commit(&message, &author, &email)
    })
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn commit_and_push(
    repo_path: String,
    message: String,
    author: String,
    email: String,
) -> CommandResult<String> {
    GitService::commit_and_push(&repo_path, &message, &author, &email).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn checkout_branch(repo_path: String, branch_name: String) -> CommandResult<()> {
    GitService::with_write_lock(&repo_path, |repo| repo.checkout_branch(&branch_name))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_branch(
    repo_path: String,
    name: String,
    target_commit_id: Option<String>,
    checkout: Option<bool>,
) -> CommandResult<()> {
    GitService::create_branch(
        &repo_path,
        &name,
        target_commit_id.as_deref(),
        checkout.unwrap_or(true),
    )
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn delete_branch(
    repo_path: String,
    name: String,
    force: Option<bool>,
) -> CommandResult<()> {
    GitService::delete_branch(&repo_path, &name, force.unwrap_or(false)).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn rename_branch(
    repo_path: String,
    old_name: String,
    new_name: String,
) -> CommandResult<()> {
    GitService::rename_branch(&repo_path, &old_name, &new_name).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_tags(repo_path: String) -> CommandResult<Vec<TagItem>> {
    GitService::open(&repo_path)
        .map_err(|e| e.to_string())?
        .list_tags()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_tag(
    repo_path: String,
    name: String,
    message: Option<String>,
    commit_id: Option<String>,
) -> CommandResult<()> {
    GitService::create_tag(&repo_path, &name, message.as_deref(), commit_id.as_deref())
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_stashes(repo_path: String) -> CommandResult<Vec<StashItem>> {
    let mut repo = GitService::open(&repo_path).map_err(|e| e.to_string())?;
    repo.list_stashes().map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn create_stash(repo_path: String, message: Option<String>) -> CommandResult<()> {
    GitService::create_stash(&repo_path, message.as_deref()).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn pop_stash(repo_path: String, index: Option<usize>) -> CommandResult<()> {
    GitService::pop_stash(&repo_path, index.unwrap_or(0)).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn reset(repo_path: String, target: String, mode: String) -> CommandResult<()> {
    GitService::reset(&repo_path, &target, &mode).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn merge(
    repo_path: String,
    target: String,
    strategy: Option<String>,
) -> CommandResult<()> {
    GitService::merge(&repo_path, &target, strategy.as_deref() == Some("no-ff"))
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn merge_abort(repo_path: String) -> CommandResult<()> {
    GitService::abort_merge(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn merge_continue(repo_path: String) -> CommandResult<String> {
    GitService::continue_merge(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cherry_pick(repo_path: String, commit_id: String) -> CommandResult<()> {
    GitService::cherry_pick(&repo_path, &commit_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cherry_pick_continue(repo_path: String) -> CommandResult<String> {
    GitService::continue_cherry_pick(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn revert(repo_path: String, commit_id: String) -> CommandResult<()> {
    GitService::revert(&repo_path, &commit_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn revert_continue(repo_path: String) -> CommandResult<String> {
    GitService::continue_revert(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_commit_changes(
    repo_path: String,
    commit_id: String,
) -> CommandResult<Vec<gitbx_core::FileStatusItem>> {
    GitService::get_commit_changes(&repo_path, &commit_id).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn fetch_remote(repo_path: String, remote_name: Option<String>) -> CommandResult<()> {
    if remote_name.is_some() {
        GitService::with_write_lock(&repo_path, |repo| {
            repo.fetch_remote(remote_name.as_deref().unwrap_or("origin"))
        })
        .map_err(|e| e.to_string())
    } else {
        GitService::fetch_all(&repo_path).map_err(|e| e.to_string())
    }
}

#[tauri::command]
pub async fn pull(repo_path: String) -> CommandResult<()> {
    GitService::pull(&repo_path, "origin").map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn push(repo_path: String) -> CommandResult<()> {
    GitService::push(&repo_path, "origin").map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn rebase(repo_path: String, upstream: String) -> CommandResult<()> {
    GitService::rebase(&repo_path, &upstream).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn rebase_continue(repo_path: String) -> CommandResult<()> {
    GitService::continue_rebase(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn operation_abort(repo_path: String) -> CommandResult<()> {
    GitService::abort_operation(&repo_path).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn worktree_add(
    repo_path: String,
    destination: String,
    branch: String,
) -> CommandResult<()> {
    GitService::worktree(&repo_path, &destination, &branch).map_err(|e| e.to_string())
}
